src/bit_machine/mod.rs

```diff
@@ -179,11 +183,14 @@ impl BitMachine {
 
     /// Add a read frame with some given value in it, as input to the
     /// program
-    pub fn input(&mut self, input: &Value) {
-        // FIXME typecheck this
+    pub fn input(&mut self, input: Arc<Value>) -> Result<(), ExecutionError> {
+        if !input.clone().is_of_type(self.source_ty.clone()) {
```

Comment:
In f45da7ac65eb7843e2a5875f3b3414298220477c

...then both these `.clone`s can go away.

Reply:
And the function can keep taking an `&Value`.